02 Charging For Your App/app.rb

```diff
@@ -0,0 +1,207 @@
+          client_secret: client_secret,
+          code: code}
+
+        response = HTTParty.post(url, body: payload)
+
+        # if the response is successful, obtain the token and store it in a hash
+        if response.code == 200
+          @tokens[shop] = response['access_token']
+        else
+          return [500, "Something went wrong."]
+        end
+
+        instantiate_session(shop)
+      end
+    end
+
+    def instantiate_session(shop)
+      # now that we have the token, we can instantiate a session
+      session = ShopifyAPI::Session.new(shop, @tokens[shop])
+      ShopifyAPI::Base.activate_session(session)
+    end
+
+    def validate_hmac(hmac,request)
+      h = request.params.reject{|k,_| k == 'hmac' || k == 'signature'}
+      query = URI.escape(h.sort.collect{|k,v| "#{k}=#{v}"}.join('&'))
+      digest = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha256'), @secret, query)
+
+      if not (hmac == digest)
+        return [403, "Authentication failed. Digest provided was: #{digest}"]
+      end
+    end
+
+    def verify_webhook(hmac, data)
+      digest = OpenSSL::Digest.new('sha256')
+      calculated_hmac = Base64.encode64(OpenSSL::HMAC.digest(digest, @secret, data)).strip
+
+      hmac == calculated_hmac
+    end
+
+    def create_recurring_application_charge
+      # checks to see if there is already an RecurringApplicationCharge created and activated
+      if not ShopifyAPI::RecurringApplicationCharge.current
+        recurring_application_charge = ShopifyAPI::RecurringApplicationCharge.new(
+                name: "Gift Basket Plan",
+                price: 4.99,
+                return_url: "https:\/\/jordo.ngrok.io\/activatecharge",
+                test: true,
+                trial_days: 7,
+                capped_amount: 100,
+                terms: "$1 for every order created")
+
+        # if the new RecurringApplicationCharge saves, we redirect to the confirmation URL,
+        # so the user can accept or decline the charge
+        if recurring_application_charge.save
+          redirect recurring_application_charge.confirmation_url
+        end
+      end
+    end
+
+    def bulk_edit_url
+      bulk_edit_url = "https://www.shopify.com/admin/bulk"\
+                    "?resource_name=ProductVariant"\
+                    "&edit=metafields.test.ingredients:string"
+      return bulk_edit_url
+    end
+
+    def create_order_webhook
+      # create webhook for order creation if it doesn't exist
+      if not ShopifyAPI::Webhook.find(:all).any?
+        webhook = {
+          topic: 'orders/create',
+          address: "https://#{@app_url}/giftbasket/webhook/order_create",
+          format: 'json'}
+
+        ShopifyAPI::Webhook.create(webhook)
+      end
+    end
+
+    def create_usage_charge
+      usage_charge = ShopifyAPI::UsageCharge.new(description: "1 dollar per order plan", price: 1.0)
+      recurring_application_charge = ShopifyAPI::RecurringApplicationCharge.last
```

Comment:
there should be a current method. (this will fail is the last recurring app charge is not active)